### PR TITLE
Add types for node's inbuilt readline/promises

### PR DESCRIPTION
**Pull request by AlekEagle**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://nodejs.org/api/readline.html#promises-api>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by AlekEagle** [comment]
my bad, forgot to add tests

File: types/node/index.d.ts [hunk 1]
```diff
@@ -40,6 +40,7 @@
 //                 NodeJS Contributors <https://github.com/NodeJS>
 //                 Linus Unnebäck <https://github.com/LinusU>
 //                 wafuwafu13 <https://github.com/wafuwafu13>
+//                 Alek Evans <https://github.com/AlekEagle>
```
Comment (amcasey) [inline types/node/index.d.ts:43]: Just confirm, this is not for allocating credit - git does that - this is a list of people who can devote time to reviewing changes to these types.  Was that your intention?

**Review by amcasey — COMMENTED** [review]
LGTM, but I'd like to hear from an owner.

**Comment by peterblazejewicz** [comment]
notice: we could move `InterfaceEvents` into `readline` itself in the future to share and compact events implementation

**Comment by peterblazejewicz** [comment]
my only concern is that `readline` module events are defined in more open way, without limiting params and even allowing to use event type not yet defined in DT, say that way:
```ts
on<K extends keyof InterfaceEvents>(event: K, listener: InterfaceEvents[K]): this;
on(event: string, listener: (...args: any[]) => void): this;
```
thoughts on this?

File: types/node/test/readline.ts [hunk 2]
```diff
@@ -52,8 +53,8 @@ const rl: readline.ReadLine = readline.createInterface(new stream.Readable());
 }
 
 {
-    rl.question("query", (answer: string) => {});
```
Comment (SimonSchick) [inline types/node/test/readline.ts:55]: Please revert unrelated style changes, they are confusing to reviewers.

File: types/node/readline/promises.d.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        SIGCONT: () => void;
+        SIGINT: () => void;
+        SIGTSTP: () => void;
+    }
+    type Completer = (line: string) => CompleterResult | Promise<CompleterResult>;
+    class Interface extends EventEmitter {
+        readonly terminal: boolean;
+        readonly line: string;
+        readonly cursor: number;
+        protected constructor(input: NodeJS.ReadableStream, output?: NodeJS.WritableStream, completer?: Completer, terminal?: boolean);
```
Comment (SimonSchick) [inline types/node/readline/promises.d.ts:23]: Should we expose those? Is `Interface` designed to be extended?

File: types/node/readline/promises.d.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        removeListener<K extends keyof InterfaceEvents>(eventName: K, listener: InterfaceEvents[K]): this;
+        removeAllListeners(eventName?: keyof InterfaceEvents): this;
+        listeners<K extends keyof InterfaceEvents>(eventName: K): Array<InterfaceEvents[K]>;
+        rawListeners<K extends keyof InterfaceEvents>(eventName: K): Array<InterfaceEvents[K]>;
+        eventNames(): Array<keyof InterfaceEvents>;
+        listenerCount(eventName: keyof InterfaceEvents): number;
+        prependListener<K extends keyof InterfaceEvents>(eventName: K, listener: InterfaceEvents[K]): this;
+        prependOnceListener<K extends keyof InterfaceEvents>(eventName: K, listener: InterfaceEvents[K]): this;
+    }
+    type ReadLine = Interface;
```
Comment (SimonSchick) [inline types/node/readline/promises.d.ts:49]: Why this alias?

File: types/node/readline/promises.d.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        line: (input: string) => void;
+        history: (history: string[]) => void;
+        pause: () => void;
+        resume: () => void;
+        SIGCONT: () => void;
+        SIGINT: () => void;
+        SIGTSTP: () => void;
+    }
+    type Completer = (line: string) => CompleterResult | Promise<CompleterResult>;
+    class Interface extends EventEmitter {
```
Comment (SimonSchick) [inline types/node/readline/promises.d.ts:19]: According to the docs this extends `readline.Interface`, you can remove most of this class declaration.

**Review by SimonSchick — CHANGES_REQUESTED** [review]

File: types/node/index.d.ts (hunk 1 above)
Comment (AlekEagle) [reply types/node/index.d.ts:43]: I thought it was just for credit, I'll be sure to remove it.

File: types/node/readline/promises.d.ts (hunk 3 above)
Comment (AlekEagle) [reply types/node/readline/promises.d.ts:23]: They are exposed in the callback API variant.

File: types/node/readline/promises.d.ts (hunk 4 above)
Comment (AlekEagle) [reply types/node/readline/promises.d.ts:49]: It was to maintain consistency with the callback API, but was only left for legacy code, so it could probably be removed.

File: types/node/test/readline.ts (hunk 2 above)
Comment (AlekEagle) [reply types/node/test/readline.ts:55]: I thought I configured my prettier to prevent this. My bad!

File: types/node/readline/promises.d.ts (hunk 4 above)
Comment (SimonSchick) [reply types/node/readline/promises.d.ts:49]: Let's remove it since this a new module.

File: types/node/readline/promises.d.ts (hunk 5 above)
Comment (AlekEagle) [reply types/node/readline/promises.d.ts:19]: I tried to extend the callback class and override the different functions, I may have done it wrong but it did not work for me.

File: types/node/test/readline.ts (hunk 2 above)
Comment (AlekEagle) [reply types/node/test/readline.ts:55]: The style change from double quotes to single was from prettier, and that is the intended setting for the repositories prettier, should I keep it or revert it anyway?

**Comment by AlekEagle** [comment]
that will allow for the declared events to be inferred while still allowing the generics. correct?

**Comment by peterblazejewicz** [comment]
> that will allow for the declared events to be inferred while still allowing the generics. correct?
> […](#)

Yes, in the essence.

**Comment by AlekEagle** [comment]
alright, I've been busy working on getting my house ready to go onto the market because I'm moving, but I'll be sure to get my latest updates pushed.

File: types/node/test/readline.ts (hunk 2 above)
Comment (AlekEagle) [reply types/node/test/readline.ts:55]: I'm bumping this as this is something I can target when working on my next commit

**Comment by AlekEagle** [comment]
I'm still working on getting ready to move, I'll probably get more work done on it when it gets closer to being stable instead of experimental.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by peterblazejewicz** [comment]
@AlekEagle I've fixed the types that were causing CI tests to fail on your origin branch. I took a liberty on co-pushing changes on properly rebased branch, so the CI tests are run against most recent master history. That was only done to not loose a work you've put into this, as this was already approved by Node maintainer. Thanks!

File: types/node/test/readline.ts (hunk 2 above)
Comment (jakebailey) [reply types/node/test/readline.ts:55]: I would revert this to get the PR moving forward. It's a shame that we have an unenforced prettier config, so until we do actually enforce it and bulk fix everything, I would just go for the minimal diff.

Comment (jakebailey) [reply types/node/test/readline.ts:55]: Oh, this is a really thread, my mistake. Seems like things have continued anyway.

**Review by mcollina — CHANGES_REQUESTED** [review]
Please revert all the unneeded style changes

**Comment by AlekEagle** [comment]
changes will be made as soon as I am finished moving, and I'll be solidifying definitions as soon as this moves out of experimental API
